distributed/shuffle/_worker_extension.py

```diff
@@ -230,8 +227,18 @@ async def get_output_partition(self, i: PartitionIDType) -> PartitionType:
         """Get an output partition to the shuffle run"""
 
 
+@dataclass(frozen=True)
+class ArrayRechunkShardID:
+    """Unique identifier of an individual shard within an array rechunking"""
+
+    #: Index of the new chunk the shard belongs
+    new_index: NIndex
+    #: Sub-index of the shard within the new chunk
+    sub_index: NIndex
```

Comment:
you might want to use slots for this. That can otherwise be pretty expensive, both in terms of runtime overhead but also in terms of memory

```suggestion
@dataclass(frozen=True)
class ArrayRechunkShardID:
    """Unique identifier of an individual shard within an array rechunking"""

    #: Index of the new chunk the shard belongs
    new_index: NIndex
    #: Sub-index of the shard within the new chunk
    sub_index: NIndex

    __slots__ = tuple(__annotations__)
```

Reply:
this is particularly important since you already mentioned running into performance problems with this, even as plain tuples